flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlIncrementalCleanup.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.state.ttl;
+
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.runtime.state.KeyedStateBackend;
+import org.apache.flink.util.CloseableIterator;
+import org.apache.flink.util.FlinkRuntimeException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnegative;
+import javax.annotation.Nonnull;
+
+import java.util.ConcurrentModificationException;
+
+/**
+ * Incremental cleanup of state with TTL.
+ *
+ * @param <K> type of state key
+ * @param <N> type of state namespace
+ */
+class TtlIncrementalCleanup<K, N> {
+	private static final Logger LOG = LoggerFactory.getLogger(TtlIncrementalCleanup.class);
+
+	@Nonnegative
+	private final int cleanupSize;
+
+	@Nonnull
+	private final KeyedStateBackend<K> keyContext;
+
+	private AbstractTtlState<K, N, ?, ?, ?> ttlState;
+	private CloseableIterator<Tuple2<N, K>> keyNamespaceIterator;
+	private boolean suppressCallback;
```

Comment:
Please add some comments, I would suggest at least on `suppressCallback` and `cleanupSize`